java-checks/src/main/java/org/sonar/java/npe/NpeVisitor.java

```diff
@@ -0,0 +1,295 @@
+        state = new NPEState.NotNull(parameter);
+      }
+      executionState.markValueAs(parameter.symbol(), state);
+    }
+  }
+
+  @Override
+  protected boolean isSymbolRelevant(Symbol symbol) {
+    //ignore unknown symbol and fields.
+    return !symbol.isUnknown() && !symbol.owner().isTypeSymbol();
+  }
+
+  @Override
+  public void visitAssignmentExpression(AssignmentExpressionTree tree) {
+    super.visitAssignmentExpression(tree);
+    Symbol symbol = getSymbol(tree.variable());
+    if (symbol != null && symbol.owner().isMethodSymbol()) {
+      setStateOfValue(symbol, tree.expression());
+    }
+  }
+
+  @Override
+  public void visitVariable(VariableTree tree) {
+    super.visitVariable(tree);
+    setStateOfValue(tree.symbol(), tree.initializer());
+  }
+
+  @Override
+  public void visitArrayAccessExpression(ArrayAccessExpressionTree tree) {
+    super.visitArrayAccessExpression(tree);
+    if (!tree.expression().is(Tree.Kind.MEMBER_SELECT) && isNullTree(tree.expression())) {
+      reportIssue(tree);
+    }
+  }
+
+  private void setStateOfValue(Symbol symbol, @Nullable ExpressionTree tree) {
+    if (tree == null || isNullTree(tree)) {
+      executionState.markValueAs(symbol, new NPEState.Null(tree));
+    } else {
+      executionState.markValueAs(symbol, new NPEState.NotNull(tree));
+    }
+  }
+
+  private boolean isNullTree(ExpressionTree tree) {
+    Symbol symbol = getSymbol(tree);
+    return (symbol != null && executionState.hasState(symbol, NPEState.Null.class)) ||
+      tree.is(Tree.Kind.NULL_LITERAL) ||
+      (tree.is(Tree.Kind.METHOD_INVOCATION) && ((MethodInvocationTree) tree).symbol().metadata().isAnnotatedWith("javax.annotation.CheckForNull"));
+  }
+
+  @Override
+  public void visitMemberSelectExpression(MemberSelectExpressionTree tree) {
+    super.visitMemberSelectExpression(tree);
+    if (isNullTree(tree.expression())) {
+      reportIssue(tree.expression());
+    }
+  }
+
+  @Override
+  public void visitMethodInvocation(MethodInvocationTree tree) {
+    super.visitMethodInvocation(tree);
+    if (tree.symbol().isMethodSymbol()) {
+      JavaSymbol.MethodJavaSymbol methodSymbol = (JavaSymbol.MethodJavaSymbol) tree.symbol();
+      List<JavaSymbol> parameters = methodSymbol.getParameters().scopeSymbols();
+      if (!parameters.isEmpty()) {
+        for (int i = 0; i < tree.arguments().size(); i += 1) {
+          if (parameters.get(i < parameters.size() ? i : parameters.size() - 1).metadata().isAnnotatedWith("javax.annotation.Nonnull")
+            && isNullTree(tree.arguments().get(i))) {
+            reportIssue(tree.arguments().get(i));
+            parameterErrorsByMethodName.put(tree.arguments().get(i), methodSymbol.name());
+          }
+        }
+      }
+    }
+  }
+
+  private void reportIssue(ExpressionTree tree) {
+    executionState.reportIssue(tree);
+    //Set value to not null to do not report multiple times
+    executionState.markValueAs(getSymbol(tree) , new NPEState.NotNull(tree));
```

Comment:
:new: ![CRITICAL](https://raw.githubusercontent.com/henryju/image-hosting/master/critical.png)Possible null pointer dereference in org.sonar.java.npe.NpeVisitor.reportIssue(ExpressionTree) due to return value of called method[![...](https://raw.githubusercontent.com/henryju/image-hosting/master/more.png)](http://nemo.sonarqube.org/coding_rules#rule_key=findbugs%3ANP_NULL_ON_SOME_PATH_FROM_RETURN_VALUE)